doc/developer/chips/18.rst

```diff
@@ -0,0 +1,825 @@
+architectures in Chapel.  These came up in conversations but for various
+reasons weren't considered at length.
+
+
+Locale per NUMA Domain
+......................
+
+Instead of running one Chapel program instance per system node, on NUMA
+architectures we could run one program instance per NUMA domain.  This
+would let us continue with the simplicity of using single-ddata with
+first-touch localization for ``DefaultRectangularArr`` arrays.  The
+downside is that we would need to go to a multi-locale domain mapping
+such as ``Block`` for arrays too large to fit on a NUMA domain, whereas
+today we do not need to do so until they are too large to fit on a
+compute node.  Thus more array elements would be remote.  The impact
+this would have on performance could be reduced by using any of several
+techniques for bypassing the network for inter-process memory references
+within a compute node.
+
+This alternative seems like it would reduce overall performance because
+it would give:
+
+* the same performance for arrays small enough to fit in a NUMA domain
+
+* decidedly less performance for arrays too large for a single NUMA
+  domain but small enough for a single locale, because of the need to
+  use the ``Block`` domain where we currently do not (assuming any of
+  the array localization techniques described above work out)
+
+* slightly less performance for arrays too large to fit in a single
+  locale, because of the involvement of the ``Block`` domain map for
+  references to other NUMA domains on the same locale
+
+
+Distribution over Sublocales
+............................
+
+The ``Block`` distribution currently places data across the predefined
+``Locales`` array, thus across the top-level network-connected compute
+nodes.  But it could be modified to place data across an array of the
+NUMA sublocales instead.  Each DefaultRectangularArr sub-array of the
+distribution would reside within a single NUMA domain, so we could use
+single-ddata with first-touch localization.  In the domain map code,
+references to sub-arrays on network-remote locales could be done as we
+currently do (using network transactions) and references to sub-arrays
+in other NUMA sublocales of the network-local locale could be done using
+regular memory references.
+
+One objection to this was that although theoretically it should be
+straightforward to modify ``Block`` to distribute over any array of
+locales rather than over the predefined ``Locales`` array, in practice
+there are likely implicit interactions with, and/or subtle dependencies
+on, the current Chapel execution model that runs a program instance per
+compute node.  If so, implementing this idea might be harder than
+expected.
+
+As with "Locale per NUMA Domain" above, this seems likely to produce
+lower performance with arrays too large for a single NUMA domain but
+small enough for a single locale, because of the added involvement of
+the ``Block`` domain map where it is not currently used.
+
```

Comment:
This is specifically talking about single-locale arrays. That might not be clear, though I thought it should be because I talked about arrays smaller than a NUMA domain, those larger than a NUMA domain but smaller than a compute node, and those larger than a compute node (all in the previous section).  In this particular case, consider a single-locale program with an array that takes up 2/3 of compute node memory (and assume 2 NUMA domains).  With any of the single-ddata alternatives, or multi-ddata, that array doesn't need the Block distribution.  You can just declare it without a domain map, and the references go directly through DefaultRectangularArr mechanisms. But with the Distribution-over-Sublocales alternative you would domain map that array using Block.  So all your references would go through the Block mechanisms, which I'd expect to slow things down a fair amount.